outer.c

```diff
@@ -183,6 +185,60 @@ static void persist_ns_files(pid_t pid, const char **persist)
 	}
 }
 
+/*
+ * If bst has entered a cgroup this function will epoll the cgroup.events file
+ * to detect when all pids have exited the cgroup ("populated 0"). The cgroup is 
+ * destroyed when this condition is met.
+ */
+static void cgroup_helper(int cleanfd, pid_t rootpid)
+{
+	char *subcgroup = makepath("bst.%d", rootpid);
+	int cgroupfd = openat(cleanfd, subcgroup, O_DIRECTORY);
+
+	int cevent = openat(cgroupfd, "cgroup.events", 0);
+	if (cevent == -1) {
+		err(1, "unable to open cgroup.events");
+	}
+	
+	/* Use EPOLLET to be notified of any changes to the cgroup.events file without
+	   needing to seek through the entire file (which seems problematic with this
+	   kernel interface).*/
+	struct epoll_event event = {
+		.events = EPOLLET,
+		.data.fd = 0
+	};
+
+	int epollfd = epoll_create1(0);
+	if (epollfd == -1) {
+		err(1, "epoll_create1");
+	}
+
+	// Notify about I/O changes to cgroupfd
+	if (epoll_ctl(epollfd, EPOLL_CTL_ADD, cevent, &event) == -1) {
+		err(1, "epoll_ctl_add cgroupfd");
+	}
+
+	char populated[BUFSIZ];
+	int ready = 0;
+
+	do {
+		// Single event (update cgroup.procs), block indefinitely
+		ready = epoll_wait(epollfd, &event, 1, -1);
+
+		// We need a new fd to read from the cgroup.events
+		int readfd = openat(cgroupfd, "cgroup.events", O_RDONLY);
+		read(readfd, &populated, BUFSIZ);
+		close(readfd);
+
+		// Check if there are no procs within the bst cgroup -- if so, delete it
+		if (strncmp(populated, "populated 0", 11) == 0) {
```

Comment:
The order in which keys appear in the file is not specified. It's more work, but I'd rather you process this file line by line, and search for `populated 0`.

fdopen, fseek, and fgets are your friends. I'd also assume that no line is larger than BUFSIZ characters -- you can error out if fgets returns sizeof (populated) and the last character is not a newline.

Reply:
Now iterating through all lines and checking for "populated 0".